modules/rte_mis_allocation/src/Controller/StudentAdmissionReportController.php

```diff
@@ -0,0 +1,721 @@
+use Drupal\user\UserInterface;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+use Symfony\Component\HttpFoundation\RedirectResponse;
+use Symfony\Component\HttpKernel\Exception\NotFoundHttpException;
+
+/**
+ * Provides role based details for student.
+ */
+final class StudentAdmissionReportController extends ControllerBase {
+
+  /**
+   * The current user service.
+   *
+   * @var \Drupal\Core\Session\AccountProxyInterface
+   */
+  protected $currentUser;
+
+  /**
+   * The entity type manager service.
+   *
+   * @var \Drupal\Core\Entity\EntityTypeManagerInterface
+   */
+  protected $entityTypeManager;
+
+  /**
+   * The config factory service.
+   *
+   * @var \Drupal\Core\Config\ConfigFactoryInterface
+   */
+  public $configFactory;
+
+  /**
+   * Constructs the controller instance.
+   */
+  public function __construct(
+    EntityTypeManagerInterface $entityTypeManager,
+    AccountProxyInterface $currentUser,
+    ConfigFactoryInterface $config_factory,
+  ) {
+    $this->entityTypeManager = $entityTypeManager;
+    $this->currentUser = $currentUser;
+    $this->configFactory = $config_factory;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container) {
+    return new static(
+      $container->get('entity_type.manager'),
+      $container->get('current_user'),
+      $container->get('config.factory'),
+    );
+  }
+
+  /**
+   * Acces for district and block admin.
+   *
+   * @param \Drupal\Core\Session\AccountInterface $account
+   *   Current user account.
+   * @param \Drupal\Core\Routing\RouteMatchInterface $routeMatch
+   *   Current routeMatch.
+   *
+   * @return \Drupal\Core\Access\AccessResultAllowed|\Drupal\Core\Access\AccessResultForbidden
+   *   If the user can access to the allotment report dashboard.
+   */
+  public function access(AccountInterface $account, RouteMatchInterface $routeMatch) {
+    // Checks if a district/block admin cannot access
+    // their adjacent or above hierarchy data.
+    $id = $routeMatch->getParameter('id') ?? NULL;
+    $currentUser = $this->entityTypeManager->getStorage('user')->load($account->id());
+    $currentUserRole = $currentUser->getRoles(TRUE);
+
+    if (array_intersect(['district_admin', 'block_admin'], $currentUserRole)) {
+      $currentUserLocation = $currentUser->get('field_location_details')->getString() ?? NULL;
+      if ($currentUserLocation) {
+        $locationTree = $this->entityTypeManager->getStorage('taxonomy_term')->loadTree('location', $currentUserLocation, NULL, FALSE);
+        if ($currentUserLocation == $id) {
+          return AccessResult::allowed()->setCacheMaxAge(0);
+        }
+        foreach ($locationTree as $value) {
+          if ($value->tid == $id) {
+            return AccessResult::allowed()->setCacheMaxAge(0);
+          }
+        }
+      }
+
+      return AccessResult::allowedIf($id == NULL)->setCacheMaxAge(0);
+    }
+    return AccessResult::allowed()->setCacheMaxAge(0);
```

Comment:
When are not returning access denied?

Reply:
we are returning access denied if a district / block, tries to access its parents id or same level location id.For example district A tries to access the data of district B. Or if a block tries to access a districts data.